### ISSUE TITLE
Snap docs/configs/image versions

### ISSUE THREAD

**Issue by mattklein123** [envoyproxy/envoy#6361]
Related to https://github.com/envoyproxy/envoy/pull/6357. We have two different user bases:
1) People that run master
2) People that run tagged versions

We need to do a better job of having a consistent state that works for people. I think this will involve setting up some templating so that when we build the docs, we snap to a specific SHA for everything, including the image tags. This will need more thought since some of our references to `latest` are checked into code (so maybe some of the Docker files also need to be generated.)

**Comment by yifan-gu** [comment]
@mattklein123 Thanks for the response on #6363 !

Yeah, was a little bit surprised when seeing the `envoyproxy/envoy:v1.9.0` tag has changed to point something that's built just 2 days ago.

IMOH, once tagged with `major.minor.patch` I won't expect it to change..

**Comment by mattklein123** [comment]
I don't think the tag has changed. What changed is `latest` no longer points to the master build but points to v1.9.0.

**Comment by bndw** [comment]
The release tag has not changed

```
docker inspect --format '{{ .Created }}' envoyproxy/envoy:v1.9.0
2018-12-20T20:40:31.342647689Z
```

**Comment by bndw** [comment]
re: templating and GITSHAs-

Would this approach work for releases, where we use the tag instead of the sha? We build on both events so it seems possible. I think maintaining friendly release tags in the release docs would go a long way for user experience.

If it was the case that the following was true:
- All commits to master have docs with an `envoyproxy/envoy-dev` image
- All releases have docs with their respective `envoyproxy/envoy:vX.X.X` image

Then I think it would solve the inconsistency problem. Then it seems like a matter of making sure folks end up on the right version of documentation and the current docs are great because they have a version selector. This may sound crazy, but what if you changed the default branch visible in Github to the latest release 🤔

**Comment by mattklein123** [comment]
@bndw we do already build tagged docs. See here: https://www.envoyproxy.io/docs

We still have more to do though, because some of those docs refer back into the repo which links to master. I think some of this is pretty easy to fix.

1) We can make the `:repo:` link use the git SHA of the doc build. I can do this today.
2) I think for anthing that refers to docker images, we need to use the SHA and not latest. This is going to take some more thought. I think we could have a macro for references in RST docs, but then we have some MD files, and we also have the Dockerfiles that are checked in. Any thoughts on this one?

**Comment by yifan-gu** [comment]
> The release tag has not changed
> 
> ```
> docker inspect --format '{{ .Created }}' envoyproxy/envoy:v1.9.0
> 2018-12-20T20:40:31.342647689Z
> ```

Ah, I was just looking at the dockerhub tag created date, thank you for the clarification! @bndw

**Comment by bndw** [comment]
> I think for anthing that refers to docker images, we need to use the SHA and not latest. This is going to take some more thought. I think we could have a macro for references in RST docs, but then we have some MD files, and we also have the Dockerfiles that are checked in. Any thoughts on this one?

I agree that pinning the Docker images is necessary and understand from a technical standpoint why using the SHA makes sense. The downside I see to this approach is that by using the SHA, correct me if I'm wrong, you'll be using `envoyproxy/envoy-dev`, even for release docs.

**Comment by mattklein123** [comment]
@bndw take a look at the linked PR, we pivot from envoy-dev to envoy if it's a tagged release.

**Comment by bndw** [comment]
Thanks. That's what I get for commenting before coffeeing.

**Comment by mattklein123** [comment]
@cpaika if you are looking for another issue related to the examples you have been working on, do you feel like finishing this one up? The remaining work here is to figure out a way that we don't refer to `:latest` in all of the Dockerfiles for the examples. We need to somehow inject the version to use via an environment variable, and then correctly reference that command in the snapped docs, as I have done for the other commands in the docs (see https://github.com/envoyproxy/envoy/pull/6376).

**Comment by cpaika** [comment]
@mattklein123 Sorry for the late response, just got over the worst cold.  Sounds perfect I'll get right on it.  Super appreciate the suggestions for next steps from everyone, really helps make it approachable to get involved as someone who hasn't contributed to open source before
